test/es12/optional-chaining.js

```diff
@@ -0,0 +1,223 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Copyright (c) ChakraCore Project Contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+// @ts-check
+/// <reference path="../UnitTestFramework/UnitTestFramework.js" />
+
+WScript.LoadScriptFile("..\\UnitTestFramework\\UnitTestFramework.js");
+
+const simpleObj = { "null": null, "undefined": undefined };
+Object.freeze(simpleObj);
+
+const tests = [
+    {
+        name: "Simple properties",
+        body() {
+            // Verify normal behavior
+            assert.throws(() => simpleObj.nothing.something, TypeError);
+            assert.throws(() => simpleObj.null.something, TypeError);
+            assert.throws(() => simpleObj.undefined.something, TypeError);
+
+            // With optional-chains
+            assert.isUndefined(simpleObj.nothing?.something, "OptChain should evaluated to 'undefined'");
+            assert.isUndefined(simpleObj.null?.something, "OptChain should evaluated to 'undefined'");
+            assert.isUndefined(simpleObj.undefined?.something, "OptChain should evaluated to 'undefined'");
+        }
+    },
+    {
+        name: "Simple indexers",
+        body() {
+            // Verify normal behavior
+            assert.throws(() => simpleObj.nothing["something"], TypeError);
+            assert.throws(() => simpleObj.null["something"], TypeError);
+            assert.throws(() => simpleObj.undefined["something"], TypeError);
+
+            // With optional-chains
+            assert.isUndefined(simpleObj.nothing?.["something"], "OptChain should evaluated to 'undefined'");
+            assert.isUndefined(simpleObj.null?.["something"], "OptChain should evaluated to 'undefined'");
+            assert.isUndefined(simpleObj.undefined?.["something"], "OptChain should evaluated to 'undefined'");
+        }
+    },
+    {
+        name: "Simple method calls",
+        body() {
+            // Verify normal behavior
+            assert.throws(() => simpleObj.nothing(), TypeError);
+            assert.throws(() => simpleObj.null(), TypeError);
+            assert.throws(() => simpleObj.undefined(), TypeError);
+
+            // With optional-chains
+            assert.isUndefined(simpleObj.nothing?.(), "OptChain should evaluated to 'undefined'");

```

Comment:
Jitted code throws a `TypeError` here (it should not)
```
TypeError: Object doesn't support property or method 'nothing'
```

Reply:
I used `LdFld` in d8841a99a5e98b39bfdfbdeb501c48acc0f3a548

Reply:
We might want to change `LdMethodFld` behavior anyway cause this bytecode
```
    0065   ProfiledLdFld        R10 = R11.nothing #3 <3>
    0069   StartCall            ArgCount: 1
```
produces a very generic error msg
```
TypeError: Object expected
```